### ISSUE TITLE
Build Error on Ubuntu 22.10: error: pointer ‘dec’ used after ‘free’ [-Werror=use-after-free]

### ISSUE THREAD

**Issue by jeremylowery** [Yubico/yubico-piv-tool#411]
I'm trying to do a built of 2.3.0 because I don't think the current stable ppa works with openssl-3.0

Steps to reproduce. (I also tried it with current HEAD).

```
jeremy@razer15:~/src$ lsb_release  -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.10
Release:        22.10
Codename:       kinetic

jeremy@razer15:~/src$ git clone git@github.com:Yubico/yubico-piv-tool.git
Cloning into 'yubico-piv-tool'...
remote: Enumerating objects: 9166, done.
remote: Counting objects: 100% (1105/1105), done.
remote: Compressing objects: 100% (441/441), done.
remote: Total 9166 (delta 686), reused 1006 (delta 618), pack-reused 8061
Receiving objects: 100% (9166/9166), 4.22 MiB | 8.76 MiB/s, done.
Resolving deltas: 100% (6610/6610), done.
jeremy@razer15:~/src$ cd yubico-piv-tool/
jeremy@razer15:~/src/yubico-piv-tool$ git checkout tags/yubico-piv-tool-2.3.0
Note: switching to 'tags/yubico-piv-tool-2.3.0'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at c7ee1b4 Updates for release 2.3.0
jeremy@razer15:~/src/yubico-piv-tool$ mkdir build
jeremy@razer15:~/src/yubico-piv-tool$ cd build
jeremy@razer15:~/src/yubico-piv-tool/build$ cmake ..
-- The C compiler identification is GNU 12.2.0
-- The CXX compiler identification is GNU 12.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.2") 
-- Performing Test HAVE_SHORTEN_64_TO_32
-- Performing Test HAVE_SHORTEN_64_TO_32 - Failed
-- Checking for module 'libcrypto'
--   Found libcrypto, version 3.0.5
        OpenSSL version:   3.0.5
-- Looking for explicit_bzero
-- Looking for explicit_bzero - found
lib/CMakeList.txt
Detected neither Mac nor Windows: selecting pcsc backend
-- Checking for module 'libpcsclite'
--   Found libpcsclite, version 1.9.9
PCSC_LIBRARIES: -L/usr/lib/x86_64-linux-gnu;-lpcsclite
PCSC_CFLAGS: -pthread -I/usr/include/PCSC
BACKEND_PCSC: ON
HAVE_PCSC_WINSCARD_H: 
lib/tests/CMakeList.txt
-- Checking for module 'check'
--   Found check, version 0.15.2
ykcs11/CMakeList.txt
ykcs11/tests/CMakeList.txt
tool/CMakeList.txt
lib/tool/CMakeList.txt
Running on Windows. Skipping basic tests...
Build summary:

        Project name:     yubico-piv-tool
        Version:          2.3.0
        Host type:        Linux
        Install prefix:   
        Compiler:         /usr/bin/cc
        Compiler ID:      GNU
        CFLAGS:            -Wall -Werror -Wno-missing-braces -Wformat -Wformat-security -Wshadow -Wmissing-prototypes -Wbad-function-cast -fstack-protector-all -Wno-pointer-sign -Wno-unused-result 
        CFLAGS_DEBUG:     -g -g2 -fno-omit-frame-pointer
        CPPFLAGS:          
        Warnings:         
        Build type:       Debug
        Backend:          check
        PCSC
                CFLAGS:   -pthread;-I/usr/include/PCSC
                LIBS:     pcsclite
        Winscard
                LIBS:     
        Mac PCSC
                LIBS:     
        Custom PCSC
                LIBS:     

        Install prefix:    /usr/local
        Install targets
                Libraries  /usr/local/lib
                Includes   /usr/local/include
                Binaries   /usr/local/bin
                Manuals    /usr/local/share/man
                Pkg-config /usr/local/lib/pkgconfig

        YKCS11 debug:     
        Hardware tests:   Disabled
-- Configuring done
-- Generating done
-- Build files have been written to: /home/jeremy/src/yubico-piv-tool/build
jeremy@razer15:~/src/yubico-piv-tool/build$ make
[  1%] Building C object lib/CMakeFiles/ykpiv.dir/ykpiv.c.o
[  3%] Building C object lib/CMakeFiles/ykpiv.dir/util.c.o
[  5%] Building C object lib/CMakeFiles/ykpiv.dir/version.c.o
[  7%] Building C object lib/CMakeFiles/ykpiv.dir/error.c.o
[  9%] Building C object lib/CMakeFiles/ykpiv.dir/internal.c.o
[ 11%] Building C object lib/CMakeFiles/ykpiv.dir/__/common/openssl-compat.c.o
[ 13%] Building C object lib/CMakeFiles/ykpiv.dir/__/common/util.c.o
[ 15%] Linking C static library libykpiv.a
[ 15%] Built target ykpiv
[ 16%] Building C object lib/CMakeFiles/ykpiv_shared.dir/ykpiv.c.o
[ 18%] Building C object lib/CMakeFiles/ykpiv_shared.dir/util.c.o
[ 20%] Building C object lib/CMakeFiles/ykpiv_shared.dir/version.c.o
[ 22%] Building C object lib/CMakeFiles/ykpiv_shared.dir/error.c.o
[ 24%] Building C object lib/CMakeFiles/ykpiv_shared.dir/internal.c.o
[ 26%] Building C object lib/CMakeFiles/ykpiv_shared.dir/__/common/openssl-compat.c.o
[ 28%] Building C object lib/CMakeFiles/ykpiv_shared.dir/__/common/util.c.o
[ 30%] Linking C shared library libykpiv.so
[ 30%] Built target ykpiv_shared
[ 32%] Building C object lib/tests/CMakeFiles/test_basic.dir/basic.c.o
[ 33%] Linking C executable test_basic
[ 33%] Built target test_basic
[ 35%] Building C object lib/tests/CMakeFiles/test_api.dir/api.c.o
[ 37%] Linking C executable test_api
[ 37%] Built target test_api
[ 39%] Building C object lib/tests/CMakeFiles/test_parse_key.dir/parse_key.c.o
[ 41%] Linking C executable test_parse_key
[ 41%] Built target test_parse_key
[ 43%] Building C object ykcs11/CMakeFiles/ykcs11.dir/ykcs11.c.o
[ 45%] Building C object ykcs11/CMakeFiles/ykcs11.dir/token.c.o
[ 47%] Building C object ykcs11/CMakeFiles/ykcs11.dir/mechanisms.c.o
[ 49%] Building C object ykcs11/CMakeFiles/ykcs11.dir/utils.c.o
[ 50%] Building C object ykcs11/CMakeFiles/ykcs11.dir/openssl_utils.c.o
[ 52%] Building C object ykcs11/CMakeFiles/ykcs11.dir/objects.c.o
[ 54%] Building C object ykcs11/CMakeFiles/ykcs11.dir/__/common/openssl-compat.c.o
[ 56%] Building C object ykcs11/CMakeFiles/ykcs11.dir/__/common/util.c.o
[ 58%] Linking C static library libykcs11.a
[ 58%] Built target ykcs11
[ 60%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/ykcs11.c.o
[ 62%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/token.c.o
[ 64%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/mechanisms.c.o
[ 66%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/utils.c.o
[ 67%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/openssl_utils.c.o
[ 69%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/objects.c.o
[ 71%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/__/common/openssl-compat.c.o
[ 73%] Building C object ykcs11/CMakeFiles/ykcs11_shared.dir/__/common/util.c.o
[ 75%] Linking C shared library libykcs11.so
[ 75%] Built target ykcs11_shared
[ 77%] Building C object ykcs11/tests/CMakeFiles/test_ykcs11.dir/ykcs11_tests.c.o
[ 79%] Building C object ykcs11/tests/CMakeFiles/test_ykcs11.dir/ykcs11_tests_util.c.o
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c: In function ‘test_rsa_decrypt’:
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:64:23: error: pointer ‘dec’ used after ‘free’ [-Werror=use-after-free]
   64 | #define asrt(c, e, m) _asrt(__FILE__, __LINE__, c, e, m);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:1236:7: note: in expansion of macro ‘asrt’
 1236 |       asrt(funcs->C_DecryptUpdate(session, enc+108, 20, dec, &dec_len), CKR_OK, "DECRYPT UPDATE");
      |       ^~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:1226:7: note: call to ‘free’ here
 1226 |       free(dec);
      |       ^~~~~~~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:64:23: error: pointer ‘dec’ used after ‘free’ [-Werror=use-after-free]
   64 | #define asrt(c, e, m) _asrt(__FILE__, __LINE__, c, e, m);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:1234:7: note: in expansion of macro ‘asrt’
 1234 |       asrt(funcs->C_DecryptUpdate(session, enc+100, 8, dec, &dec_len), CKR_OK, "DECRYPT UPDATE");
      |       ^~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:1226:7: note: call to ‘free’ here
 1226 |       free(dec);
      |       ^~~~~~~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:64:23: error: pointer ‘dec’ used after ‘free’ [-Werror=use-after-free]
   64 | #define asrt(c, e, m) _asrt(__FILE__, __LINE__, c, e, m);
      |                       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:1232:7: note: in expansion of macro ‘asrt’
 1232 |       asrt(funcs->C_DecryptUpdate(session, enc, 100, dec, &dec_len), CKR_OK, "DECRYPT UPDATE");
      |       ^~~~
/home/jeremy/src/yubico-piv-tool/ykcs11/tests/ykcs11_tests_util.c:1226:7: note: call to ‘free’ here
 1226 |       free(dec);
      |       ^~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [ykcs11/tests/CMakeFiles/test_ykcs11.dir/build.make:90: ykcs11/tests/CMakeFiles/test_ykcs11.dir/ykcs11_tests_util.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:380: ykcs11/tests/CMakeFiles/test_ykcs11.dir/all] Error 2
make: *** [Makefile:146: all] Error 2


```

**Comment by jeremylowery** [comment]
For what it's worth if I delete the function test_rsa_decrypt from the test file then it compiles clean. Assume it's localized to that function.

**Comment by qpernil** [comment]
Weird those line numbers don't match up even though I'm checked out to the same commit. Im on Ubuntu 22.04 which uses gcc GNU 11.3.0, I'll assume that's why. But looking at the test_rsa_decrypt function I have to say it looks correct to me. Could you try just using separate variables for each allocation ? If that works I'll change the code, even though I think it's correct as is..

**Comment by qpernil** [comment]
I apologize, I somehow looked at the wrong version of the code. You are correct that there was a bug in that version, it was fixed in the commit directly after. In fact I'd recommend building the latest master, which is closer to the next release which will come some time after the new year.

**Comment by qpernil** [comment]
In fact, there was another fix (commit 720afc81f5e034d2a8d3944e4f28463f808998a4) for a use-after free warning from GCC 12 after that initial fix for the bug. So just taking the next commit would still fail I believe, given that warnings are treated as errors. Again, please use master for now!

**Comment by qpernil** [comment]
I have confirmed the above on my own Ubuntu 22.10 installation now, so I'll close the issue. Please feel free to reopen if needed.
